presto-spark-base/src/main/java/com/facebook/presto/spark/execution/PrestoSparkRowBuffer.java

```diff
@@ -63,21 +65,34 @@ public void setNoMoreRows()
         }
     }
 
-    public PrestoSparkRow get()
+    public List<PrestoSparkRow> get()
             throws InterruptedException
     {
-        PrestoSparkRow row = null;
+        List<PrestoSparkRow> rows = null;
         synchronized (monitor) {
             while (buffer.isEmpty() && !finished) {
                 monitor.wait();
             }
             if (!buffer.isEmpty()) {
-                row = buffer.poll();
+                rows = buffer.poll();
             }
-            if (row != null) {
-                memoryManager.updateMemoryUsage(-row.getRetainedSize());
+            if (rows != null) {
+                memoryManager.updateMemoryUsage(-getRetainedSize(rows));
             }
         }
-        return row;
+        return rows;
+    }
+
+    /**
+     * Does not use iterators / streams for efficiency
+     */
+    @SuppressWarnings("ForLoopReplaceableByForEach")
+    private static long getRetainedSize(List<PrestoSparkRow> rows)
```

Comment:
Is this only get called twice ? (once `enqueue` and once `get`)? Note in theory we can always reduce the retain size computation to be once (by wrapping it with something like `PrestoSparkRowBatch`. But if it's just computed twice I think it's OK to keep it as is.

Reply:
I think it makes sense. Currently the retained memory is recomputed several times. Added `BufferedRows` class to reuse the `currentBatchSize` computed by the `PrestoSparkOutputOperator`